Assets/SEE/UI/Window/ConsoleWindow/ConsoleWindow.cs

```diff
@@ -0,0 +1,691 @@
+    /// <para>
+    ///     The console messages can be categorized by channels and levels:
+    ///     <list type="bullet">
+    ///         <item><description>Each channel has a fontawesome icon, which is displayed in front of each console message.</description></item>
+    ///         <item><description>Each level (in every channel) specifies the color which is used for the message backgrounds.</description></item>
+    ///         <item><description>The channels and levels be used to filter specific messages.</description></item>
+    ///         <item><description>By default only the channel `User Input` with one level `Log` exists.</description></item>
+    ///     </list>
+    /// </para>
+    /// <seealso cref="Channel"/>
+    /// <seealso cref="ChannelLevel"/>
+    /// </summary>
+    public class ConsoleWindow : BaseWindow
+    {
+        /// <summary>
+        /// The user submitted the input.
+        /// </summary>
+        public static event UnityAction<string> OnInputSubmit;
+
+        /// <summary>
+        /// The user changed the input.
+        /// </summary>
+        public static event UnityAction<string> OnInputChanged;
+
+        /// <summary>
+        /// The window prefab.
+        /// </summary>
+        private const string windowPrefab = "Prefabs/UI/ConsoleWindow/ConsoleWindow";
+        /// <summary>
+        /// The prefab for each message.
+        /// </summary>
+        private const string itemPrefab = "Prefabs/UI/ConsoleWindow/ConsoleWindowItem";
+        /// <summary>
+        /// The number of spaces to use for tabs.
+        /// <seealso cref="tabReplacement"/>.
+        /// </summary>
+        private const int tabSize = 4;
+
+        /// <summary>
+        /// The replacement for tabs in console messages.
+        /// Tabs are replaced with spaces for two reasons:
+        /// (1) TeshMeshPro doesn't work well with tabs. (weird spacing)
+        /// (2) The search field doesn't allow tabs.
+        /// <seealso cref="tabSize"/>
+        /// </summary>
+        private static readonly string tabReplacement = new(' ', tabSize);
+
+        /// <summary>
+        /// All console messages.
+        /// </summary>
+        private static List<Message> messages = new List<Message>();
+        /// <summary>
+        /// The messages got cleared.
+        /// </summary>
+        private static event Action MessagesCleared;
+        /// <summary>
+        /// A message was added.
+        /// </summary>
+        private static event Action MessageAdded;
+        
+        /// <summary>
+        /// A message was changed.
+        /// </summary>
+        private static event Action MessageChanged;
+
+        /// <summary>
+        /// A channel or level was changed.
+        /// </summary>
+        private static event Action ChannelChanged;
+
+        /// <summary>
+        /// The messages were cleared.
+        /// </summary>
+        private bool messagesCleared;
```

Comment:
If I remember correctly, this was done because instantiating and destroying game objects is only allowed on the main thread. (so doing that directly inside the UI-event isn't allowed)

Reply:
Since the actions are executed on `Update()` (which is executed on the main thread), that shouldn't be a problem anymore, right? Maybe it's left over from a previous approach?

Reply:
I am not quite sure what you mean?
The `messagesCleared`, `messageAdded`, etc. ensure that each frame only one thing is done.

Otherwise deleting and then adding new messages would cause the new messages also being deleted.
(since Unity doesn't destroy game objects immediately, except if `DestroyImmediately` is used...)

Reply:
Maybe I understand you wrong, but it is necessary to keep track of both separately.

It is possible that first the console is cleared and then a message is added (on the same frame).
Then the console should be cleared first, and the message added the frame after that.

How would I keep track of that with only two properties?

Reply:
Sorry, I should have been more specific. We could use the `[Flags]` property on the enum here. You can take a look at the top of `LSPImporter.cs` (in most recent master) for an example. An event can then be set using e.g. `lastEvent |= ConsoleEvent.Cleared` (without invalidating any previously set values).